### ISSUE TITLE
[css-text-decor] Should text-decoration-skip-ink skip parts within in glyphs?

### ISSUE THREAD

**Issue by SebastianZ** [w3c/csswg-drafts#4504]
The current implementations of `text-decoration-skip-ink` in Firefox and Safari skip over the inside of glyphs leaving gaps within the underlines.

@litherum [mentioned in #4493](https://github.com/w3c/csswg-drafts/issues/4493#issuecomment-551257774) that he implemented it this way because "little bits of underlines inside bowls of letters like lowercase g [...] were too ugly to ship." and in https://bugzil.la/1575383 the opinions for having underlines displayed within glyphs were ambivalent.

[Text Decoration 3](https://drafts.csswg.org/css-text-decor-3/#line-decoration) currently has a note about this saying:

> Note, this specification intentionally does not mandate a particular method for “following the shape” of the glyph so that UAs can take appropriate measures to handle aesthetic and performance considerations.

I believe in some cases it does make sense to show the underlines in the inner parts of the glyphs, especially in combination with negative `text-underline-offset`s and/or bigger font sizes.

Example:

Showing underline within glyphs:
![underlining without skipping inner parts of glyphs](https://user-images.githubusercontent.com/958943/68546371-cec29f00-03d5-11ea-9975-52b023a78b62.png)

Skipping underline within characters:
![underlining with skipping inner parts of glyphs](https://user-images.githubusercontent.com/958943/68546396-00d40100-03d6-11ea-94c3-4f43c3cf2e1f.png)

See https://jsfiddle.net/SebastianZ/qprzf15b/ or for browsers not supporting `text-underline-offset` yet https://jsfiddle.net/SebastianZ/pe8rf4a3/.

Sebastian

**Comment by jfkthame** [comment]
It's not a general solution for all cases, obviously, but for simple examples an effect like this could be "faked" by turning off the current skip-ink behavior, and then stroking the text with the background color; try https://codepen.io/jfkthame/pen/mddjVXp in Firefox.

Whether this is worth browsers implementing "properly" as a built-in skip-ink behavior seems a bit questionable to me; I'm having a hard time thinking of a case where using a large font-size and negative underline-offset would be sensible design! (I do have more sympathy for the case of Cyrillic Д, and no doubt there are other characters with similar issues.)

**Comment by fantasai** [comment]
Personally I'd go with something like

![skip-ink](https://user-images.githubusercontent.com/725717/68554561-d8500500-03dd-11ea-89e1-29646a12ebbf.png)

which has a vertical-axis check for "internal" rather than a horizontal one, and matches the line edges to the glyph shapes.

But this is an area where the spec isn't mandating things intentionally. If you don't like how a particular implementation is doing it, file bugs on them for poor quality of implementation. :)

**Comment by SebastianZ** [comment]
> But this is an area where the spec isn't mandating things intentionally. If you don't like how a particular implementation is doing it, file bugs on them for poor quality of implementation. :)

That's what I did initially. See https://bugzil.la/1575383, which I already linked above.

The spec. doesn't need to mandate things but it could at least describe a preference for what should be done. Or at least mention that font size and line thickness may also have an influence on the skipping.

And yes, your suggestion may also be a solution.

Sebastian
